Subject: Make xmlError structs constant
From: Michal Privoznik mprivozn@redhat.com Mon Nov 27 15:04:43 2023 +0100
Date: Mon Nov 27 15:04:43 2023 +0100:
Git: 34378a4ac257f2f5fcf364786d1634a8c36b304f

In libxml2 commits v2.12.0~14 and v2.12.0~77 the API changed so
that:

1) xmlGetLastError() returns pointer to a constant xmlError
   struct, and

2) xmlSetStructuredErrorFunc() changed the signature of callback
   (validate_structured_error_nop()), it too is passed pointer to
   a constant xmlError struct.

But of course, older libxml2 expects different callback
signature. Therefore, we need to typecast it anyway.

Also, drop obviously incorrect @error annotation in
validate_structured_error_nop; the variable is used.

Signed-off-by: Michal Privoznik <mprivozn@redhat.com>

diff --git a/tools/osinfo-db-validate.c b/tools/osinfo-db-validate.c
index a721b4d..b1434a6 100644
--- a/tools/osinfo-db-validate.c
+++ b/tools/osinfo-db-validate.c
@@ -35,7 +35,7 @@ static void validate_generic_error_nop(void *userData G_GNUC_UNUSED,
 }
 
 static void validate_structured_error_nop(void *userData G_GNUC_UNUSED,
-                                          xmlErrorPtr error G_GNUC_UNUSED)
+                                          const xmlError *error)
 {
     if (error->file)
         g_printerr("%s:%d %s", error->file, error->line, error->message);
@@ -173,7 +173,8 @@ static gboolean validate_files(GFile *schema, gsize nfiles, GFile **files, GErro
     g_autofree gchar *schemapath = NULL;
 
     xmlSetGenericErrorFunc(NULL, validate_generic_error_nop);
-    xmlSetStructuredErrorFunc(NULL, validate_structured_error_nop);
+    /* Drop this typecast when >=libxml2-2.12.0 is required */
+    xmlSetStructuredErrorFunc(NULL, (xmlStructuredErrorFunc) validate_structured_error_nop);
 
     schemapath = g_file_get_path(schema);
     rngParser = xmlRelaxNGNewParserCtxt(schemapath);
